[Tag] Adds support for url prop

jbalsas opened the pull request:
This is a spin off of [[WIP] [Tag] Allow arbitrary content and allow onClick and onRemove at the same time](https://github.com/Shopify/polaris-react/pull/4830). I'm opening it early as a Draft so I can get feedback while I start working on tests and final cleanup 🙏 

This PR adds a `url` prop to the `Tag` component that works in the following way:
- You can either pass `onClick` or `url` but not both (this is enforced in the type system)
- You can pass both `url` and `onRemove`
- When `url` is received, an anchor tag is used to render the tag content
- When both `url` and `onRemove` are provided, a `segmented` variant of the tag is created and a `ButtonGroup` is used to render both items

### WHY are these changes introduced?

This is a partial approach to solve the problem presented in [[Tag] Allow arbitrary children](https://github.com/Shopify/polaris-react/issues/4819) by focusing on some part of the specific need that triggered that request (navigation in a removable tag).

I've limited this PR to just the interaction, since arbitrary children sounds like a less controversial topic and switched gears based on @dleroux [comment](https://github.com/Shopify/polaris-react/pull/4830#issuecomment-993916486):

> What do you think about leaving the onClick and onRemove as is, but adding a url prop?

### WHAT is this pull request doing?

Adds the following options that can be explored in the Storybook `URL Tag` section:

```jsx
// Tags with links
<Tag url="https://www.shopify.com/">Wholesale</Tag>
<Tag url="https://www.shopify.com/">Long text that causes ellipsis</Tag>
<Tag url="https://www.shopify.com/" onRemove={() => {}}>Wholesale</Tag>
<Tag url="https://www.shopify.com/" onRemove={() => {}}>Long text that causes ellipsis</Tag>

// Tags with custom content
<Tag>
  <Stack alignment="center" spacing="extraTight" title="Order #123">
    <Icon source={SearchMinor} />
    <span>Order #123</span>
    <Badge status="attention" progress="partiallyComplete">
      In Progress
    </Badge>
  </Stack>
</Tag>
```

https://user-images.githubusercontent.com/905006/146237385-a5ec3609-7365-4b3b-a725-c1bd3326fc14.mov

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

Look into the Storybook `URL Tag` section:

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
- [ ] For visual design changes, ping @ sarahill to update the Polaris UI kit

jbalsas committed:
[Tag] Adds support for url prop

This commit adds a `url` prop to the `Tag` component that works in the
following way:
- You can either pass `onClick` or `url` but not both (this is enforced
in the type system)
- You can pass both `url` and `onRemove`
- When `url` is received, an anchor tag is used to render the tag content
- When both `url` and `onRemove` are provided, a `segmented` variant of
the tag is created and both items are rendered

jbalsas commented on src/components/Tag/Tag.scss:
```diff
@@ -52,6 +48,20 @@ $icon-size: rem(16px);
+
+  &.segmented {
+    > * {
+      max-width: 100%;
```
This is so ellipsis will work in the segmented case, otherwise I think the inline element has no bound size and overflow won't kick in

jbalsas commented on src/components/Tag/Tag.scss:
```diff
@@ -52,6 +48,20 @@ $icon-size: rem(16px);
+      max-width: 100%;
+
+      > * {
+        max-width: 97%;
```
This is so that the content leaves some room for remove button when the it requires ellipsis... maybe this can be done in a better way?

jbalsas commented on src/components/Tag/Tag.tsx:
```diff
@@ -17,24 +19,46 @@ export interface NonMutuallyExclusiveProps {
+    onRemove && url && styles.segmented,
   );
 
+  if (onClick) {
```
This is all basically some refactoring that should keep the old options the same and only change something when `url` is set.

Since `onClick` is the most restrictive prop and using a ternary for this case was complex, I decided to return early here, but this could be moved around in any way in case there are some specific conventions I'm not aware of.

jbalsas committed:
[Tag] Accepts ReactElement to be passed as children

jbalsas committed:
[Tag] Updates UNRELEASED.md

chloerice commented on src/components/Tag/README.md:
````diff
@@ -80,6 +80,101 @@ Use to allow merchants to add attributes to an object.
 <Tag onClick={() => console.log('Clicked')}>Wholesale</Tag>
 ```
 
+### Tagss with links
````
```suggestion
### Tag with link
```

chloerice commented on src/components/Tag/README.md:
```diff
@@ -80,6 +80,101 @@ Use to allow merchants to add attributes to an object.
 
+### Tagss with links
+
+Use to allow merchants to navigate to an entity.
```
```suggestion
Use to allow merchants to navigate to a resource, like a customer segment or a smart collection.
```

AndrewMusgrave replied:
```suggestion
Use to allow merchants to navigate to a resource. For example a customer segment or a smart collection
```

Building on Chloe's suggestion a little

chloerice commented on src/components/Tag/README.md:
```diff
@@ -80,6 +80,101 @@ Use to allow merchants to add attributes to an object.
+        rhoncus massa. Vivamus lectus nunc, fringilla euismod risus commodo,
+        mattis blandit nulla
+      </Tag>
+    </Stack>
```
```suggestion
    <Tag url="/collections/wholesale">Wholesale</Tag>;
```

chloerice commented on src/components/Tag/README.md:
```diff
@@ -80,6 +80,101 @@ Use to allow merchants to add attributes to an object.
+
+### Tag with custom content
+
+Use to allow merchants to generate special custom content for a tag.
```
```suggestion
Use when a tag needs to be visually distinguished from others, like when it's added automatically.
```

chloerice commented on src/components/Tag/README.md:
```diff
@@ -80,6 +80,101 @@ Use to allow merchants to add attributes to an object.
+      </Badge>
+    </Stack>
+  </Tag>
+</Stack>
```
We try to keep the examples simple and reference real use cases when possible, so when `WandMinor` ships in `@shopify/polaris-icons` we should update this to reflect the real admin use case. We def want to test worst case/edge case scenarios, but we can do this in the playground 💯  

```suggestion
<Tag url="/customers?tag=VIP">
  <Stack spacing="extraTight">
    <Icon source={CustomersMinor} />
    <span>VIP</span>
  </Stack>
</Tag>
```

chloerice commented on src/components/Tag/Tag.tsx:
```diff
@@ -10,40 +10,73 @@ import styles from './Tag.scss';
+  /** A string to use when tag has more than textual content */
+  title?: string;
+  /** Url to navigate to when tag is clicked or keypressed. */
+  url?: string;
```
Let's give the `title` prop a descriptive name of `accessibilityLabel` (this also aligns with with naming in other components, like Button etc).

```suggestion
  children?: string | React.ReactElement;
  /** Disables the tag  */
  disabled?: boolean;
    /** Descriptive text to provide context for screen reader users. */
  accessibilityLabel?: string;
  /** Url to navigate to when tag is clicked or keypressed. */
  url?: string;
  /** Callback when tag is clicked or keypressed. Renders without remove button or url when set. */
  onClick?(): void;
  /** Callback when remove button is clicked or keypressed. */
  onRemove?(): void;
```

chloerice commented on src/components/Tag/Tag.tsx:
```diff
@@ -10,40 +10,73 @@ import styles from './Tag.scss';
+    );
+  }
+
+  let tagTitle = title;
```
```suggestion
  let tagTitle = accessibilityLabel;
```

chloerice commented on src/components/Tag/Tag.tsx:
```diff
@@ -10,40 +10,73 @@ import styles from './Tag.scss';
+  disabled = false,
+  onClick,
+  onRemove,
+  title = '',
```
```suggestion
  accessibilityLabel = '',
```

AndrewMusgrave replied:
Should we leave out a default value? Right now we can render a title without an content, but if we left it as undefined react would omit it

jbalsas replied:
Hey @AndrewMusgrave, I tried this (turns out already tried it before) and the reason I left it as empty was because down there we have

```
const ariaLabel = i18n.translate('Polaris.Tag.ariaLabel', {
  children: tagTitle,
});
```

So, `children` expects `string | number`. If `title` can be `undefined`, then we'd need additional logic for that as well, which might be fine... what do you think?

chloerice commented on src/components/Tag/tests/Tag.test.tsx:
```diff
@@ -35,4 +37,78 @@ describe('<Tag />', () => {
+    });
+  });
+
+  describe('title', () => {
```
```suggestion
  describe('accessibilityLabel', () => {
```

chloerice commented on src/components/Tag/tests/Tag.test.tsx:
```diff
@@ -35,4 +37,78 @@ describe('<Tag />', () => {
+      });
+    });
+
+    it('uses the title prop to render title when a string is received as children', () => {
```
```suggestion
    it('uses the accessibilityLabel prop to render title when a string is received as children', () => {
```

chloerice commented on src/components/Tag/tests/Tag.test.tsx:
```diff
@@ -35,4 +37,78 @@ describe('<Tag />', () => {
+      });
+    });
+
+    it('uses the title prop to render title when a component is received as children and url prop is passed', () => {
```
```suggestion
    it('uses the accessibilityLabel prop to render title when a component is received as children and url prop is passed', () => {
```

chloerice reviewed (approved):
This works great @jbalsas!! 

The only problem I'm seeing in my playground is with the edge case of a `Tag` with really long content, the margin of the `Stack` causes  the `Tag` to overflow. But this also happens even on the `main` branch so it's not related to your changes.

<details>
<summary>Click to view collapsed screenshot 👀 </summary>

<img width="1041" alt="Screen Shot 2021-12-15 at 2 25 44 PM" src="https://user-images.githubusercontent.com/18447883/146251915-786745f0-237c-4e06-970a-a0592b4ed4da.png">

</details>

<details>
<summary>Click to view collapsed playground code 👀 </summary>

```jsx
import React from 'react';
import {CustomersMinor, OrderStatusMinor} from '@shopify/polaris-icons';

import {Card, Stack, Badge, Page, Tag, Icon, Layout} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      <Layout>
        <Layout.Section secondary>
          <Card sectioned>
            <Stack>
              <Tag url="/customers?tag=VIP" onRemove={() => {}}>
                <Stack wrap={false} spacing="extraTight">
                  <Icon source={CustomersMinor} />
                  <span>VIP</span>
                </Stack>
              </Tag>
              <Tag
                url="/customers?tag=VIP"
                onRemove={() => {}}
                accessibilityLabel="VIP"
              >
                <Stack alignment="center" wrap={false} spacing="extraTight">
                  <Icon source={OrderStatusMinor} />
                  <span>3PL</span>
                  <span style={{position: 'relative', top: '-1px'}}>
                    <Badge status="attention" progress="partiallyComplete">
                      Partially fulfilled
                    </Badge>
                  </span>
                </Stack>
              </Tag>
              <Tag>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer
                at ipsum quam. Aliquam fermentum bibendum vestibulum. Vestibulum
                condimentum luctus metus, sed sagittis magna pellentesque eget.
                Duis dapibus pretium nisi, et venenatis tortor dignissim ut.
                Quisque eget lacus ac ex eleifend ultrices. Phasellus facilisis
                ex sit amet leo elementum condimentum. Ut vel maximus felis.
                Etiam eget diam eu eros blandit interdum. Sed eu metus sed justo
                aliquam iaculis ac sit amet ex. Curabitur justo magna, porttitor
                non pulvinar eu, malesuada at leo. Cras mollis consectetur eros,
                quis maximus lorem dignissim at. Proin in rhoncus massa. Vivamus
                lectus nunc, fringilla euismod risus commodo, mattis blandit
                nulla.
              </Tag>
              <Tag url="/customers?tag=VIP" onRemove={() => {}}>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer
                at ipsum quam. Aliquam fermentum bibendum vestibulum. Vestibulum
                condimentum luctus metus, sed sagittis magna pellentesque eget.
                Duis dapibus pretium nisi, et venenatis tortor dignissim ut.
                Quisque eget lacus ac ex eleifend ultrices. Phasellus facilisis
                ex sit amet leo elementum condimentum. Ut vel maximus felis.
                Etiam eget diam eu eros blandit interdum. Sed eu metus sed justo
                aliquam iaculis ac sit amet ex. Curabitur justo magna, porttitor
                non pulvinar eu, malesuada at leo. Cras mollis consectetur eros,
                quis maximus lorem dignissim at. Proin in rhoncus massa. Vivamus
                lectus nunc, fringilla euismod risus commodo, mattis blandit
                nulla.
              </Tag>
              <Tag url="/customers?tag=VIP">
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer
                at ipsum quam. Aliquam fermentum bibendum vestibulum. Vestibulum
                condimentum luctus metus, sed sagittis magna pellentesque eget.
                Duis dapibus pretium nisi, et venenatis tortor dignissim ut.
                Quisque eget lacus ac ex eleifend ultrices. Phasellus facilisis
                ex sit amet leo elementum condimentum. Ut vel maximus felis.
                Etiam eget diam eu eros blandit interdum. Sed eu metus sed justo
                aliquam iaculis ac sit amet ex. Curabitur justo magna, porttitor
                non pulvinar eu, malesuada at leo. Cras mollis consectetur eros,
                quis maximus lorem dignissim at. Proin in rhoncus massa. Vivamus
                lectus nunc, fringilla euismod risus commodo, mattis blandit
                nulla.
              </Tag>{' '}
            </Stack>
          </Card>
        </Layout.Section>
      </Layout>
    </Page>
  );
}
```

</details>

AndrewMusgrave commented on src/components/Tag/Tag.scss:
```diff
@@ -96,4 +98,44 @@ $icon-size: rem(16px);
+  color: var(--p-text);
+  outline: none;
+  border-radius: var(--p-border-radius-base);
+  text-decoration: none;
```
Links normally have underlines. Right now the button tag/link tag seem almost visually identical. Should we add underlines to differentiate them?

jbalsas replied:
My initial approach had indeed underlines, but @ouellettejordan asked to remove them. I agree that it's not obvious that those actions are going to be navigations... 

Maybe one compromise would be to add the [ExternalMinor](https://polaris-icons.shopify.com/?icon=ExternalSmallMinor&q=exter) icon at the end to signal navigation, but this is usually done for external navigations and nothing guarantees that's the case here.

How big of a deal this is @AndrewMusgrave? A quick search for `text-decoration: none` in `web` returns **`80 results - 68 files`** so I think that cat is out of the bag... 🐱 

That being said, there's no occurrence of `text-decoration: none` in here, so I can see how we might want to think this through and see if we should expose this differently.

chloerice replied:
When to underline has always been a contentious thing 😅 But sometimes what we want/like and what merchants need in order to have clear expectations is not the same 🥲 I'd recommend adding `text-decoration` back when hovered or focused (that's the compromise we landed on for the plain `Button` variant).

dleroux replied:
> underline has always been a contentious thing 😅

😅

I do like the visual cue that the text is actually a link on focus of hover.

AndrewMusgrave commented on src/components/Tag/Tag.scss:
```diff
@@ -96,4 +98,44 @@ $icon-size: rem(16px);
+    @include focus-ring($style: 'focused');
+  }
+
+  &:hover {
```
Do we need a disabled / active state? And can we have Button/Link share the same styles so it's easier to update in the future?

jbalsas replied:
From what I saw, each case has some differences with the other. 

There's some common code that at some point I tried to hide away in a `interaction()` mixin. I can give that a try again. My understanding of what everything is doing is pretty limited, so I can't promise I'll get it right though! 😂 

![dontknowwhatimdoing](https://user-images.githubusercontent.com/905006/146263114-961964ab-64ae-4ee8-8bb9-c45b1143c5f6.gif)

(Yes, regardless of what dhh says 😂 )

jbalsas replied:
> Do we need a disabled / active state?

I think we do, nice catch!!

AndrewMusgrave commented on src/components/Tag/Tag.tsx:
```diff
@@ -10,40 +10,73 @@ import styles from './Tag.scss';
 export interface NonMutuallyExclusiveProps {
   /** Content to display in the tag */
-  children?: string;
+  children?: string | React.ReactElement;
```
```suggestion
  children?: React.ReactNode;
```

`ReactNode` is the type we generally go for to allow custom markup in Polaris components. And string will also be included in `ReactNode`!

String is included in React

jbalsas replied:
Sweet!!!

AndrewMusgrave reviewed (commented):
Looking good!

jbalsas committed:
Apply suggestions from code review

Co-authored-by: Chloe Rice <chloerice@users.noreply.github.com>
Co-authored-by: Andrew Musgrave <andrew.musgrave@shopify.com>

jbalsas committed:
Polaris icons v4.11.0

jbalsas committed:
[Tag] Update custom content example with real example

jbalsas committed:
[Tag] Improve disabled states for new url and segmented options

Additionally, it fixes a possible existing issue when a clickable tag was
disabled which wasn't getting the proper disabled text color.

Use this Plaground example to test the different options

```
import React from 'react';

import {Card, Stack, Page, Tag, Layout} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      <Layout>
        <Layout.Section secondary>
          <Card sectioned>
            <Stack>
              <Tag>Normal</Tag>
              <Tag disabled>Disabled</Tag>
              <Tag onClick={() => {}}>onClick</Tag>
              <Tag disabled onClick={() => {}}>
                disabled onClick
              </Tag>
              <Tag onRemove={() => {}}>onRemove</Tag>
              <Tag disabled onRemove={() => {}}>
                disabled onRemove
              </Tag>
              <Tag url="#">url</Tag>
              <Tag disabled url="#">
                disabled url
              </Tag>
              <Tag url="#" onRemove={() => {}}>
                url onRemove
              </Tag>
              <Tag disabled url="#" onRemove={() => {}}>
                disabled url onRemove
              </Tag>{' '}
            </Stack>
          </Card>
        </Layout.Section>
      </Layout>
    </Page>
  );
}
```

jbalsas committed:
[Tag] Adds underline decoration to links upon hover/focus interactions

kyledurand commented on src/components/Tag/Tag.scss:
```diff
@@ -49,9 +45,19 @@ $icon-size: rem(16px);
+  }
+
+  &.linkable {
+    align-items: stretch;
```
This jumps out as a bit of an odd property to me. Can you share some info around why you need this?

jbalsas replied:
That's a great observation!! 

I went back to the playground and I don't think that's making a difference at all! 😱 

I know at some point the `a` elements were shorter in height than the tag so hovering created a weird effect where only the inner span was being highlighted. Using stretch allowed for the whole area to be covered without having to resort to layout tricks (padding/margin/position) that made it also hard to get the focus ring right. 

At that time I was using both `UnstyledLink` and `Stack`, which I ended up removing, so maybe that played a particular role in creating that issue!

Seeing as this is no longer needed, I'll remove it!! 🎩

kyledurand commented on src/components/Tag/Tag.scss:
```diff
@@ -96,4 +98,46 @@ $icon-size: rem(16px);
+}
+
+.Link {
+  display: inline-grid;
```
This seems like it should be a flex layout to me. Curious to know more about this choice as well though

jbalsas replied:
This is the one way I found to make it work. I think the Grid layout is more strict in terms of size and does the job better in this case. I only did a shallow search for references in the specs, so don't have any concrete information as to why this is happening.

When trying to use flexbox I'd do:

```css
display: inline-flex;
max-width: 100%;
```

However, this would cause the remove button to overflow the element. 

![Screenshot 2021-12-16 at 16 10 23](https://user-images.githubusercontent.com/905006/146407361-cb21479c-771c-40ac-855d-887409639d35.png)

I tried different combinations, so I might've also missed a way to get this to work using flexbox. I'm open to suggestions here! ☺️

kyledurand replied:
Ah strange, well if this works it's better than flex hacks for sure 👍

kyledurand commented on src/components/Tag/Tag.scss:
```diff
@@ -96,4 +98,46 @@ $icon-size: rem(16px);
+  .LinkText {
+    overflow: hidden;
+    white-space: nowrap;
+    text-overflow: ellipsis;
```
You can use:
```suggestion
    @include truncate;
```

I'm not sure this is actually working as you intended though:
![image](https://screenshot.click/16-34-z88bc-xw691.png)

jbalsas replied:
>  @include truncate;

Nice, will do that!

> I'm not sure this is actually working as you intended though:

Interesting, what browser would that be? Do you have the code for the playground?

I've been testing on Chrome, Firefox and Safari and that seems to be working fine in all of them

<details><summary>This is the Playground I'm using as provided by @chloerice, maybe we need to add something else?</summary>

```jsx
import React from 'react';
import {CustomersMinor, OrderStatusMinor} from '@shopify/polaris-icons';

import {Card, Stack, Badge, Page, Tag, Icon, Layout} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      <Layout>
        <Layout.Section secondary>
          <Card sectioned>
            <Stack>
              <Tag url="/customers?tag=VIP" onRemove={() => {}}>
                <Stack wrap={false} spacing="extraTight">
                  <Icon source={CustomersMinor} />
                  <span>VIP</span>
                </Stack>
              </Tag>
              <Tag
                url="/customers?tag=VIP"
                onRemove={() => {}}
                accessibilityLabel="VIP"
              >
                <Stack alignment="center" wrap={false} spacing="extraTight">
                  <Icon source={OrderStatusMinor} />
                  <span>3PL</span>
                  <span style={{position: 'relative', top: '-1px'}}>
                    <Badge status="attention" progress="partiallyComplete">
                      Partially fulfilled
                    </Badge>
                  </span>
                </Stack>
              </Tag>
              <Tag>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer
                at ipsum quam. Aliquam fermentum bibendum vestibulum. Vestibulum
                condimentum luctus metus, sed sagittis magna pellentesque eget.
                Duis dapibus pretium nisi, et venenatis tortor dignissim ut.
                Quisque eget lacus ac ex eleifend ultrices. Phasellus facilisis
                ex sit amet leo elementum condimentum. Ut vel maximus felis.
                Etiam eget diam eu eros blandit interdum. Sed eu metus sed justo
                aliquam iaculis ac sit amet ex. Curabitur justo magna, porttitor
                non pulvinar eu, malesuada at leo. Cras mollis consectetur eros,
                quis maximus lorem dignissim at. Proin in rhoncus massa. Vivamus
                lectus nunc, fringilla euismod risus commodo, mattis blandit
                nulla.
              </Tag>
              <Tag url="/customers?tag=VIP" onRemove={() => {}}>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer
                at ipsum quam. Aliquam fermentum bibendum vestibulum. Vestibulum
                condimentum luctus metus, sed sagittis magna pellentesque eget.
                Duis dapibus pretium nisi, et venenatis tortor dignissim ut.
                Quisque eget lacus ac ex eleifend ultrices. Phasellus facilisis
                ex sit amet leo elementum condimentum. Ut vel maximus felis.
                Etiam eget diam eu eros blandit interdum. Sed eu metus sed justo
                aliquam iaculis ac sit amet ex. Curabitur justo magna, porttitor
                non pulvinar eu, malesuada at leo. Cras mollis consectetur eros,
                quis maximus lorem dignissim at. Proin in rhoncus massa. Vivamus
                lectus nunc, fringilla euismod risus commodo, mattis blandit
                nulla.
              </Tag>
              <Tag url="/customers?tag=VIP">
                Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer
                at ipsum quam. Aliquam fermentum bibendum vestibulum. Vestibulum
                condimentum luctus metus, sed sagittis magna pellentesque eget.
                Duis dapibus pretium nisi, et venenatis tortor dignissim ut.
                Quisque eget lacus ac ex eleifend ultrices. Phasellus facilisis
                ex sit amet leo elementum condimentum. Ut vel maximus felis.
                Etiam eget diam eu eros blandit interdum. Sed eu metus sed justo
                aliquam iaculis ac sit amet ex. Curabitur justo magna, porttitor
                non pulvinar eu, malesuada at leo. Cras mollis consectetur eros,
                quis maximus lorem dignissim at. Proin in rhoncus massa. Vivamus
                lectus nunc, fringilla euismod risus commodo, mattis blandit
                nulla.
              </Tag>{' '}
            </Stack>
          </Card>
        </Layout.Section>
      </Layout>
    </Page>
  );
}
```

 </details>

<details><summary>Screenshot of the Playground</summary>

![Screenshot 2021-12-16 at 15 57 06](https://user-images.githubusercontent.com/905006/146405234-afa3280c-b5c3-4b82-ae07-9ca1db98f2c1.png)

</details>

jbalsas replied:
I think if you're using `<Stack>` you'll need to also do `wrap={false}` or if you're not, you would take care of making that custom content inline, if that's what you were looking to do from a user of `Tag`.

But yeah, looks like if you have custom content, the ellipsis won't have an effect and I'm not sure if we can get it to work... shouldn't be the one crafting the custom content the one responsible for it? Should we add this as a known limitation?

kyledurand replied:
Ok cool this looks great 👍. I think it was the stack issue in my quick playground that got me

kyledurand commented on src/components/Tag/Tag.tsx:
```diff
@@ -10,40 +10,73 @@ import styles from './Tag.scss';
+}: TagProps) {
   const i18n = useI18n();
 
+  const segmented = onRemove && url;
```
This seems like it could cause bugs in web. Can't we have segmented tags without urls?

jbalsas replied:
Unless I missed something (which could very easily happen), currently we cannot. The component [prevents us from passing both `onClick` and `onRemove`](https://github.com/Shopify/polaris-react/blob/main/src/components/Tag/Tag.tsx#L22-L26). I'm not aware of the reasons for such decision and I couldn't find any literature or reference to it.

My initial attempt was to allow both `onClick` and `onRemove` at the same time to create the `segmented` option as you can see in [[WIP] [Tag] Allow arbitrary content and allow onClick and onRemove at the same time](https://github.com/Shopify/polaris-react/pull/4830)

To simplify the combinations I kept `onClick` separate since it could be weird to have both an `onClick` and a `url` (that should navigate on click) at the same time. Both scenarios (`onClick+onRemove` or `onClick+url`) have valid use cases, so we could work on allowing them if we want.

kyledurand replied:
Ah I see what you mean. I think I just thought this line was doing more than it actually was. Once I 🎩 ed it in the playground it seems pretty much fine, and not a breaking change. I did notice some weirdness around styles not being added: [screencast](https://screenshot.click/16-59-ly7cm-pm84g.mp4)

<details>
  <summary>Here's my playground</summary>

```jsx
import React from 'react';

import {Page, Stack, Tag} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      <Stack>
        <Tag onRemove={() => {}} url="#">
          tag with url
        </Tag>
        <Tag onRemove={() => {}}>tag with just onRemove</Tag>
      </Stack>
    </Page>
  );
}
```
</details>

jbalsas replied:
> I did notice some weirdness around styles not being added

Weird!! I'll investigate!!

kyledurand commented on src/components/Tag/tests/Tag.test.tsx:
```diff
@@ -35,4 +37,85 @@ describe('<Tag />', () => {
+      expect(tag).toContainReactComponent('a', {href: '#'});
+    });
+
+    it('renders plain text when the tag is disabled', () => {
```
Should we also have a test that asserts a span is rendered instead of anchor when url and disabled are passed in?

jbalsas replied:
Yeah, not sure why I didn't add it there, would it be okay to just have both assertions or do we have some rules about that and we'd rather keep them separate?

kyledurand replied:
I'm a fan of doing two assertions for a test like that but we have discrepancies throughout the repo. Feel free to do whichever

kyledurand reviewed (commented):
This is looking good just a few observations and questions but we should be able to get this merged today

jbalsas committed:
Apply suggestions from code review

This commit includes the following adjustments:
- Remove unnecessary `align-items: stretch;` rule in `linkable` items
- Use mixin `truncate` instead of manually reimplementing it here
- Flattens right border radius on segmented urls on hover
- Makes `accessibilityLabel` default to `undefined` so no `title` prop is
rendered when ignored. We still render the `Remove ` label on the button
for clarity on the action even though we can't add more information.
- Additional tests and assertions

kyledurand reviewed (approved):
🎉
